spec/models/user_spec.rb

```diff
@@ -1875,6 +1875,44 @@ def hash(password, salt)
     it "raises an error when passwords are too long" do
       expect { hash(too_long, "gravy") }.to raise_error(StandardError)
     end
+
+    it "uses the target algorithm for new users" do
+      expect(user.password_algorithm).to eq(User::TARGET_PASSWORD_ALGORITHM)
+    end
+
+    it "can use an older algorithm to verify existing passwords, then upgrade" do
+      old_algorithm = "$pbkdf2-sha256$i=5,l=32$"
+      expect(old_algorithm).not_to eq(User::TARGET_PASSWORD_ALGORITHM)
+
+      password = "poutine"
+      old_hash = hash(password, user.salt, old_algorithm)
+
+      user.update!(password_algorithm: old_algorithm, password_hash: old_hash)
+
+      expect(user.password_algorithm).to eq(old_algorithm)
+      expect(user.password_hash).to eq(old_hash)
```

Comment:
```suggestion
```

Minor but these two lines are not necessary since we're using `update!` above which raises an error if the update operation fails.

Reply:
There are things which could cause the values to change (e.g. a method defined on `User#password_hash` which intercepts the call, a database trigger, etc.) so I'd prefer to keep the extra check - it's basically free.